### ISSUE TITLE
 false positive for unused method using method references in java 8 

### ISSUE THREAD

**Issue by SRINIVASMACHERLA** [spotbugs/spotbugs#735]
Actual issue reported here - https://sourceforge.net/p/findbugs/bugs/1370/ which was already closed. Reopening bug - https://sourceforge.net/p/findbugs/bugs/1370/.  

For below scenario, findbug reports UPM_UNCALLED_PRIVATE_METHOD.

    public List<String> readLocations(final List<String> inputs) {
        return inputs.stream()
                .map(MyClass::createPath)
                // here comes more code..;
    }

    private Path createPath(final String input) {
        // doesn't matter
        return createPath(input, 0);
    }

private Path createPath(final String input, final int d) {
        // doesn't matter
        return null;
    }

I'm able to unblock my self changing createPath(final String input) method name. Is this a known issue?

**Comment by arlindohall** [comment]
I have this same issue. I was able to replicate with the following:

```java
package check;

import java.util.Arrays;
import java.util.List;
import java.util.Objects;

public class FindBugs {
    public void verify() {
        if (allNotNull()) {
            System.out.println("Hello");
        } else {
            System.out.println("Goodbye");
        }
    }

    private boolean allNotNull() {
        final List<String> strings = Arrays.asList("String1", "String2");
        return strings.stream().allMatch(this::allNotNull);
    }

    // Gives UPM_UNCALLED_PRIVATE_METHOD: Private method is never called
    private boolean allNotNull(final String status) {
        return Objects.nonNull(status);
    }
}
```

The following does not produce any bugs.
```java
package check;

import java.util.Arrays;
import java.util.List;
import java.util.Objects;

public class NoBugs {
    public boolean allNotNull() {
        final List<String> strings = Arrays.asList("String1", "String2");
        return strings.stream().allMatch(this::allNotNull);
    }

    private boolean allNotNull(final String status) {
        return Objects.nonNull(status);
    }
}
```

**Comment by SRINIVASMACHERLA** [comment]
@arlindohall Thanks for the response.

The only difference in latter code snippet is - access specifier (one with public other is private). In my case, the two methods scope must be private.

It does not looks good to change access specifier to fix this issue. Rather I would prefer different name for time being.

Hello findbugs team,

Are you fixing this issue?

**Comment by mebigfatguy** [comment]
The problem is that UPM doesn't know anything about BootstrapMethods.. I'll put a MR together in a bit.

**Comment by mebigfatguy** [comment]
So attached is a patch to fix the problem. However, it appears this is already addressed in spotbugs, a different way. so closing.

[UPM.txt](https://github.com/spotbugs/spotbugs/files/2313007/UPM.txt)
